### PR TITLE
Replace MappedFieldType#extractTerm with local query visitor

### PR DESCRIPTION
**Pull request by javanna**
The only usage of `MappedFieldType#extractTerm` comes from `SpanTermQueryBuilder` which attempts to extract a single term from a generic Query obtained from calling `MappedFieldType#termQuery`. We can move this logic directly within its only caller, and instead of using instanceof checks, we can rely on the query visitor API.

This additionally allows us to remove one of the leftover usages of TermInSetQuery#getTermData which is deprecated in Lucene

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/SpanTermQueryBuilder.java [hunk 1]
```diff
@@ -78,7 +83,25 @@ protected SpanQuery doToQuery(SearchExecutionContext context) throws IOException
             term = new Term(fieldName, BytesRefs.toBytesRef(value));
         } else {
             Query termQuery = mapper.termQuery(value, context);
-            term = MappedFieldType.extractTerm(termQuery);
+            List<Term> termsList = new ArrayList<>();
+            termQuery.visit(new QueryVisitor() {
+                @Override
+                public QueryVisitor getSubVisitor(BooleanClause.Occur occur, Query parent) {
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/SpanTermQueryBuilder.java:89]: Why do we need this override this way? Is it because of potentially having to unwrap BoostQuery around a TermQuery? I'm trying to understand why we only descend on MUST and FILTER clauses.

File: server/src/main/java/org/elasticsearch/index/query/SpanTermQueryBuilder.java [hunk 2]
```diff
@@ -78,7 +83,25 @@ protected SpanQuery doToQuery(SearchExecutionContext context) throws IOException
...
+                    }
+                    return EMPTY_VISITOR;
+                }
+
+                @Override
+                public void consumeTerms(Query query, Term... terms) {
+                    termsList.addAll(Arrays.asList(terms));
+                }
+            });
+            if (termsList.size() != 1) {
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/SpanTermQueryBuilder.java:101]: So this is because we only didn't error on TermQuery instances which are sure to have only one term, and TIS queries that have on term, in the original method? Just to confirm I understand this right.

**Review by cbuescher — COMMENTED** [review]
Looks good, I just want to make sure I understand the changes right, that's why I left two questions.

File: server/src/main/java/org/elasticsearch/index/query/SpanTermQueryBuilder.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/query/SpanTermQueryBuilder.java:94]: good catch. This is to reproduce the previous boost query unwrapping. I am trying to figure out how to repro that so that I can add a unit test. Do you have ideas? When would we get a BoostQuery or anything else that wraps a term query out of calling mapper.termQuery ?  if it was a problem for boost query, why wasn't it for e.g. ConstantScoreQuery? Either we were not covering enough queries with the previous instance of checks, or the check for boost query was not needed, or? :)

File: server/src/main/java/org/elasticsearch/index/query/SpanTermQueryBuilder.java (hunk 2 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/query/SpanTermQueryBuilder.java:106]: I added a comment. I came to the conclusion that this is not needed. I don't know how you can get either 0 or more than one term out of mapper.termQuery. But I no longer get why we had a special case for TermInSetQuery there either. How can you get that query again out of mapper.termQuery ? Did you see test fail when we commented that out in the lucene 10 branch?

So far, I've seen one legit build failure, we were hitting the previous "not a term query" error on match_only_text fields, and tests were expecting a 400 error without checking the error message. In reality, a term can totally be extracted out of those queries, but they return a ConstantScoreQuery which we don't unwrap, hence it seems we were lucky to return the expected status code for a situation that down the line would have raised an IllegalStateException. 

I added a specific check above to verity that there are positions and throw a specific error for that case.

File: server/src/main/java/org/elasticsearch/index/query/SpanTermQueryBuilder.java (hunk 1 above)
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/query/SpanTermQueryBuilder.java:94]: I thought having a query boost would wrap every other query? At least that's what AbstractQueryBuilder#toQuery() looks like, so setting a boost on the test query should work I think.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/query/SpanTermQueryBuilder.java:94]: right, I see your memories of the query refactoring are still alive, mine are gone :) I added a test around that, at least we know it worked before and it still works now. The query visitor is much more general which is probably overkill, but I am reluctant to check query types in there, I think the check for number of terms at the end provides the safety we need.

File: qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/upgrades/QueryBuilderBWCIT.java [hunk 3]
```diff
@@ -154,29 +154,29 @@ public QueryBuilderBWCIT(@Name("cluster") FullClusterRestartUpgradeStatus upgrad
...
             """
-                "span_near": {"clauses": [{ "span_term": { "keyword_field": "value1" }}, \
-                { "span_term": { "keyword_field": "value2" }}], "slop": 2, "in_order": false}
+                "span_near": {"clauses": [{ "span_term": { "text_field": "value1" }}, \
+                { "span_term": { "text_field": "value2" }}], "slop": 2, "in_order": false}
                 """,
-            new SpanNearQueryBuilder(new SpanTermQueryBuilder("keyword_field", "value1"), 2).addClause(
-                new SpanTermQueryBuilder("keyword_field", "value2")
+            new SpanNearQueryBuilder(new SpanTermQueryBuilder("text_field", "value1"), 2).addClause(
+                new SpanTermQueryBuilder("text_field", "value2")
```
Comment (javanna) [inline qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/upgrades/QueryBuilderBWCIT.java:179]: These queries were storing a span query against a field that did not have positions stored. With my change, that now causes an error in SpanQueryBuilder#toQuery and fails terms extraction for that query when we attempt to store it. I feel like that's the correct behaviour though. One weirdness here is also that despite the lack of positions, the query would work fine because it looks like the memory index used by percolator makes positions available anyways. I don't think we should make weird exception for the percolator though, I think that this scenario should have caused an error. I updated the test to perform span queries against a text field instead.

File: server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java [hunk 4]
```diff
@@ -561,29 +557,6 @@ protected void checkNoTimeZone(@Nullable ZoneId timeZone) {
...
-        if (termQuery instanceof TermInSetQuery tisQuery) {
-            PrefixCodedTerms terms = tisQuery.getTermData();
-            if (terms.size() == 1) {
-                TermIterator it = terms.iterator();
-                BytesRef term = it.next();
-                return new Term(it.field(), term);
-            }
-        }
-        if (termQuery instanceof TermQuery == false) {
-            throw new IllegalArgumentException("Cannot extract a term from a query of type " + termQuery.getClass() + ": " + termQuery);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:582]: this was previously causing 400 , catching a scenario where match_only_text, which does not store positions, is targeted by a span query. It falls here only because the mapper returns a constant score query which does not get unwrapped. We should rather make sure that there are positions stored for the field that is targeted, or we may get an IllegalStateException from WeightSpan in Lucene which causes a 500 status code.

File: qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/upgrades/QueryBuilderBWCIT.java [hunk 5]
```diff
@@ -93,8 +93,8 @@ public QueryBuilderBWCIT(@Name("cluster") FullClusterRestartUpgradeStatus upgrad
 
     static {
         addCandidate("""
-            "match": { "text_field": "value"}
-            """, new MatchQueryBuilder("text_field", "value"));
+            "match": { "keyword_field": "value"}
+            """, new MatchQueryBuilder("keyword_field", "value"));
```
Comment (cbuescher) [inline qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/upgrades/QueryBuilderBWCIT.java:97]: Did you change this so we use "keyword_field" anywhere? Doesn't seem related to the span query stuff otherwise or am I missunderstanding anything?

**Review by cbuescher — APPROVED** [review]
LGTM once the remaining tests pass, but the argument in the last test change makes sense to me. Left one minor question for my understanding.

File: qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/upgrades/QueryBuilderBWCIT.java (hunk 5 above)
Comment (javanna) [reply qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/upgrades/QueryBuilderBWCIT.java:97]: I just wanted to keep at least a query against a keyword field, to make sure its terms extraction works. I am assuming that was the purpose of having keyword, text and long field in this test.

Comment (javanna) [reply qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/upgrades/QueryBuilderBWCIT.java]: That ended up causing problems due to different parsing on different versions, I opted for always targeting the text field and removing the keyword field from the mappings.

**Comment by javanna** [comment]
The packaging tests have failed with a known, unrelated failure:  https://github.com/elastic/elasticsearch/issues/113219 .
